service/history/execution/mutable_state_builder.go

```diff
@@ -1661,21 +1661,22 @@ func (e *mutableStateBuilder) addWorkflowExecutionStartedEventForContinueAsNew(
 		ContinueAsNewInitiator:          attributes.Initiator,
 		FirstDecisionTaskBackoffSeconds: attributes.BackoffStartIntervalInSeconds,
 	}
+
+	// if ContinueAsNew as Cron or decider, recalculate the expiration timestamp and set attempts to 0
+	req.Attempt = 0
+	if attributes.RetryPolicy != nil && attributes.RetryPolicy.GetExpirationIntervalInSeconds() > 0 {
+		// expirationTime calculates from first decision task schedule to the end of the workflow
+		expirationInSeconds := attributes.RetryPolicy.GetExpirationIntervalInSeconds() + req.GetFirstDecisionTaskBackoffSeconds()
+		expirationTime := e.timeSource.Now().Add(time.Second * time.Duration(expirationInSeconds))
+		req.ExpirationTimestamp = common.Int64Ptr(expirationTime.UnixNano())
+	}
```

Comment:
Is the logic the same as the old one? There's no behavior change and it's just a code refactoring?

Reply:
yup, this is just refactoring